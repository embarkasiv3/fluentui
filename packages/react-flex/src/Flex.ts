export * from './components/Flex/index';
