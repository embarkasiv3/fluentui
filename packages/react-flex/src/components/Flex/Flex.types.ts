import * as React from 'react';
import { ComponentProps } from '@fluentui/react-utilities';

/**
 * {@docCategory Flex }
 */
export interface FlexProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  // TODO Add props and slots here
}

/**
 * {@docCategory Flex }
 */
export interface FlexState extends FlexProps {
  /**
   * Ref to the root slot
   */
  ref: React.RefObject<HTMLElement>;
}
