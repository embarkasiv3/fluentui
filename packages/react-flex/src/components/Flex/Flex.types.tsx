export interface FlexProps {
  vertical: boolean;
}
