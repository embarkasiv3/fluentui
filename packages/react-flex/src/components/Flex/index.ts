export * from './Flex';
export * from './Flex.types';
export * from './renderFlex';
export * from './useFlex';
export * from './useFlexStyles';
