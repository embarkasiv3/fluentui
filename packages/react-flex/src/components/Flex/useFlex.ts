import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { FlexProps, FlexState } from './Flex.types';

export const flexShorthandProps: (keyof FlexProps)[] = [];

const mergeProps = makeMergeProps<FlexState>({ deepMerge: flexShorthandProps });

/**
 * Create the state required to render Flex.
 *
 * The returned state can be modified with hooks such as useFlexStyles,
 * before being passed to renderFlex.
 *
 * @param props - props from this instance of Flex
 * @param ref - reference to root HTMLElement of Flex
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Flex }
 */
export const useFlex = (props: FlexProps, ref: React.Ref<HTMLElement>, defaultProps?: FlexProps): FlexState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
    },
    defaultProps,
    resolveShorthandProps(props, flexShorthandProps),
  );

  return state;
};
