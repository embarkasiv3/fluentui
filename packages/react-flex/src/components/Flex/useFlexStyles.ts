import { makeStyles, ax } from '@fluentui/react-make-styles';
import { FlexState } from './Flex.types';

/**
 * Styles for the root slot
 */
const useRootStyles = makeStyles<FlexState>([
  [
    null,
    theme => ({
      // TODO Add default styles
    }),
  ],
  // TODO Add styles conditioned on FlexState props, for example:
  // [
  //   s => s.exampleProp === 'exampleValue',
  //   theme => ({
  //
  //   }),
  // ],
]);

// TODO Add styles for any other slots, for example:
// const useExampleSlotStyles = makeStyles<FlexState>([
//   [
//     null,
//     theme => ({
//
//     }),
//   ],
// ]);

/**
 * Apply styling to the Flex slots based on the state
 * {@docCategory Flex }
 */
export const useFlexStyles = (state: FlexState): FlexState => {
  state.className = ax(useRootStyles(state), state.className);

  // TODO Hook up slot styles, for example:
  // const exampleSlotClassName = useExampleSlotStyles(state);
  // if (state.exampleSlot) {
  //   state.exampleSlot.className = ax(exampleSlotClassName, state.exampleSlot.className);
  // }

  return state;
};
