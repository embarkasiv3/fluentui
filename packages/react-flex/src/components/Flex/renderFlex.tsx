import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { FlexState } from './Flex.types';
import { flexShorthandProps } from './useFlex';

/**
 * Render the final JSX of Flex
 * {@docCategory Flex }
 */
export const renderFlex = (state: FlexState) => {
  const { slots, slotProps } = getSlots(state, flexShorthandProps);

  return (
    <slots.root {...slotProps.root}>
      {/* TODO Add additional slots in the appropritate place */}
      {state.children}
    </slots.root>
  );
};
