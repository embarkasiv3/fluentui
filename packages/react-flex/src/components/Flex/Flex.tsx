import * as React from 'react';
import { useFlex } from './useFlex';
import { FlexProps } from './Flex.types';
import { renderFlex } from './renderFlex';
import { useFlexStyles } from './useFlexStyles';

/**
 * TODO: Component docs
 * {@docCategory Flex }
 */
export const Flex: React.FunctionComponent<FlexProps & React.RefAttributes<HTMLElement>> = React.forwardRef<
  HTMLElement,
  FlexProps
>((props, ref) => {
  const state = useFlex(props, ref);

  useFlexStyles(state);
  return renderFlex(state);
});

Flex.displayName = 'Flex';
