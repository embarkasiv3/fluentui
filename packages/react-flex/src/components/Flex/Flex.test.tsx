import * as React from 'react';
import { Flex } from './Flex';
import * as renderer from 'react-test-renderer';
import { isConformant } from '../../common/isConformant';

describe('Flex', () => {
  isConformant({
    Component: Flex,
    displayName: 'Flex',
  });

  /**
   * Note: see more visual regression tests for Flex in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Flex>Default Flex</Flex>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
