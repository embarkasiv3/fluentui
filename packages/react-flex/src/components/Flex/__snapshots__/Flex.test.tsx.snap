// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders a default state 1`] = `
<div
  className=""
>
  Default Flex
</div>
`;
